File: app/src/main/java/com/salino/sali/data/repository/ShoppingRepositoryImpl.kt
package com.salino.sali.data.repository

import com.google.firebase.Timestamp
import com.google.firebase.firestore.ListenerRegistration
import com.salino.sali.data.local.entity.SyncTargetType
import com.salino.sali.data.local.source.ActivityLocalDataSource
import com.salino.sali.data.local.source.RecurringLocalDataSource
import com.salino.sali.data.local.source.ShoppingLocalDataSource
import com.salino.sali.data.model.ActivityLog
import com.salino.sali.data.model.ActivityType
import com.salino.sali.data.model.ItemStatus
import com.salino.sali.data.model.ShoppingItem
import com.salino.sali.data.remote.source.ShoppingRemoteDataSource
import com.salino.sali.data.sync.SyncQueueProcessor
import com.salino.sali.di.IoDispatcher
import com.salino.sali.domain.repository.AuthRepository
import com.salino.sali.domain.repository.ShoppingRepository
import com.salino.sali.util.normalizeItemName
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import java.util.Date
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ShoppingRepositoryImpl @Inject constructor(
    private val shoppingLocalDataSource: ShoppingLocalDataSource,
    private val shoppingRemoteDataSource: ShoppingRemoteDataSource,
    private val activityLocalDataSource: ActivityLocalDataSource,
    private val recurringLocalDataSource: RecurringLocalDataSource,
    private val authRepository: AuthRepository,
    private val syncQueueProcessor: SyncQueueProcessor,
    @IoDispatcher ioDispatcher: CoroutineDispatcher
) : ShoppingRepository {

    private val scope = CoroutineScope(SupervisorJob() + ioDispatcher)
    private val itemListeners = mutableMapOf<String, ListenerRegistration>()

    override fun observeActiveItems(householdId: String): Flow<List<ShoppingItem>> {
        ensureRemoteSync(householdId)
        return shoppingLocalDataSource.observeActiveItems(householdId)
    }

    override fun observeBoughtItems(householdId: String): Flow<List<ShoppingItem>> {
        ensureRemoteSync(householdId)
        return shoppingLocalDataSource.observeBoughtItems(householdId)
    }

    override fun observeAllItems(householdId: String): Flow<List<ShoppingItem>> {
        ensureRemoteSync(householdId)
        return shoppingLocalDataSource.observeAllItems(householdId)
    }

    override suspend fun addItem(householdId: String, item: ShoppingItem): Result<String> = runCatching {
        val now = Timestamp.now()
        val newItem = item.copy(
            id = item.id.ifBlank { UUID.randomUUID().toString() },
            normalizedName = normalizeItemName(item.name),
            status = ItemStatus.ACTIVE.name,
            createdAt = item.createdAt ?: now,
            updatedAt = now
        )
        shoppingLocalDataSource.upsertItem(householdId, newItem)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ITEM, newItem.id)

        val activity = createActivityLog(
            householdId = householdId,
            type = ActivityType.ITEM_ADDED,
            item = newItem,
            actorUserId = newItem.addedBy,
            actorDisplayName = newItem.addedByName
        )
        activityLocalDataSource.upsertActivity(activity)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ACTIVITY, activity.id)
        syncQueueProcessor.flush(householdId)
        newItem.id
    }

    override suspend fun updateItem(householdId: String, item: ShoppingItem): Result<Unit> = runCatching {
        val updatedItem = item.copy(
            normalizedName = normalizeItemName(item.name),
            updatedAt = Timestamp.now()
        )
        shoppingLocalDataSource.upsertItem(householdId, updatedItem)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ITEM, updatedItem.id)

        val actor = authRepository.observeCurrentUser().first()
        val activity = createActivityLog(
            householdId = householdId,
            type = ActivityType.ITEM_UPDATED,
            item = updatedItem,
            actorUserId = actor?.id ?: updatedItem.addedBy,
            actorDisplayName = actor?.displayName ?: updatedItem.addedByName
        )
        activityLocalDataSource.upsertActivity(activity)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ACTIVITY, activity.id)
        syncQueueProcessor.flush(householdId)
    }

    override suspend fun markAsBought(
        householdId: String,
        itemId: String,
        userId: String,
        userName: String
    ): Result<Unit> = runCatching {
        val currentItem = getItem(householdId, itemId).getOrThrow()
        val boughtAt = System.currentTimeMillis()
        val updatedItem = currentItem.copy(
            status = ItemStatus.BOUGHT.name,
            boughtBy = userId,
            boughtByName = userName,
            updatedAt = Timestamp(Date(boughtAt))
        )
        shoppingLocalDataSource.upsertItem(householdId, updatedItem)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ITEM, updatedItem.id)

        recurringLocalDataSource.findByNormalizedName(householdId, updatedItem.normalizedName)
            ?.let { recurring ->
                val updatedRecurring = recurring.copy(
                    lastCompletedAt = Timestamp(Date(boughtAt)),
                    nextDueAt = Timestamp(Date(boughtAt + recurring.intervalDays * DAY_MS)),
                    updatedAt = Timestamp.now()
                )
                recurringLocalDataSource.upsertRecurringItem(updatedRecurring)
                syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.RECURRING, updatedRecurring.id)
            }

        val activity = createActivityLog(
            householdId = householdId,
            type = ActivityType.ITEM_BOUGHT,
            item = updatedItem,
            actorUserId = userId,
            actorDisplayName = userName
        )
        activityLocalDataSource.upsertActivity(activity)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ACTIVITY, activity.id)
        syncQueueProcessor.flush(householdId)
    }

    override suspend fun markAsActive(householdId: String, itemId: String): Result<Unit> = runCatching {
        val currentItem = getItem(householdId, itemId).getOrThrow()
        val actor = authRepository.observeCurrentUser().first()
        val updatedItem = currentItem.copy(
            status = ItemStatus.ACTIVE.name,
            boughtBy = null,
            boughtByName = null,
            updatedAt = Timestamp.now()
        )
        shoppingLocalDataSource.upsertItem(householdId, updatedItem)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ITEM, updatedItem.id)

        val activity = createActivityLog(
            householdId = householdId,
            type = ActivityType.ITEM_RESTORED,
            item = updatedItem,
            actorUserId = actor?.id.orEmpty(),
            actorDisplayName = actor?.displayName.orEmpty()
        )
        activityLocalDataSource.upsertActivity(activity)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ACTIVITY, activity.id)
        syncQueueProcessor.flush(householdId)
    }

    override suspend fun deleteItem(householdId: String, itemId: String): Result<Unit> = runCatching {
        val currentItem = shoppingLocalDataSource.getItem(householdId, itemId)
        val actor = authRepository.observeCurrentUser().first()
        shoppingLocalDataSource.deleteItem(householdId, itemId)
        syncQueueProcessor.enqueueDelete(householdId, SyncTargetType.ITEM, itemId)

        if (currentItem != null) {
            val activity = createActivityLog(
                householdId = householdId,
                type = ActivityType.ITEM_DELETED,
                item = currentItem,
                actorUserId = actor?.id.orEmpty(),
                actorDisplayName = actor?.displayName.orEmpty()
            )
            activityLocalDataSource.upsertActivity(activity)
            syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ACTIVITY, activity.id)
        }

        syncQueueProcessor.flush(householdId)
    }

    override suspend fun getItem(householdId: String, itemId: String): Result<ShoppingItem> = runCatching {
        shoppingLocalDataSource.getItem(householdId, itemId)
            ?: shoppingRemoteDataSource.getItem(householdId, itemId).also {
                shoppingLocalDataSource.upsertItem(householdId, it)
            }
    }

    override suspend fun flushPendingSync(householdId: String): Result<Unit> = runCatching {
        syncQueueProcessor.flush(householdId)
    }

    private fun ensureRemoteSync(householdId: String) {
        if (itemListeners.containsKey(householdId)) return
        registerListener(householdId)
    }

    override fun forceRefreshSync(householdId: String) {
        // Remove old listener (may be stale after backgrounding)
        itemListeners.remove(householdId)?.remove()
        registerListener(householdId)
    }

    override fun clearListeners() {
        itemListeners.values.forEach { it.remove() }
        itemListeners.clear()
    }

    private fun registerListener(householdId: String) {
        itemListeners[householdId] = shoppingRemoteDataSource.listenToItems(householdId) { items ->
            scope.launch {
                shoppingLocalDataSource.mergeRemoteItems(householdId, items)
                syncQueueProcessor.flush(householdId)
            }
        }
    }

    private fun createActivityLog(
        householdId: String,
        type: ActivityType,
        item: ShoppingItem,
        actorUserId: String,
        actorDisplayName: String
    ): ActivityLog = ActivityLog(
        id = UUID.randomUUID().toString(),
        householdId = householdId,
        type = type.name,
        itemId = item.id,
        itemName = item.name,
        actorUserId = actorUserId,
        actorDisplayName = actorDisplayName,
        createdAt = Timestamp.now()
    )

    companion object {
        private const val DAY_MS = 24 * 60 * 60 * 1000L
    }
}

File: ios/Data/Repositories/ShoppingRepositoryImpl.swift
import Foundation
import FirebaseFirestore

@MainActor
final class ShoppingRepositoryImpl: ShoppingRepository {
    private let localStore: LocalStore
    private let remoteDataSource: ShoppingRemoteDataSource
    private let authRepository: AuthRepository
    private let syncQueueProcessor: SyncQueueProcessor
    private var itemListeners: [String: ListenerRegistration] = [:]

    init(
        localStore: LocalStore,
        remoteDataSource: ShoppingRemoteDataSource,
        authRepository: AuthRepository,
        syncQueueProcessor: SyncQueueProcessor
    ) {
        self.localStore = localStore
        self.remoteDataSource = remoteDataSource
        self.authRepository = authRepository
        self.syncQueueProcessor = syncQueueProcessor
    }

    func observeActiveItems(householdId: String) -> AsyncStream<[ShoppingItem]> {
        ensureRemoteSync(householdId)
        return localStore.observeShoppingItems(householdId: householdId, status: .active)
    }

    func observeBoughtItems(householdId: String) -> AsyncStream<[ShoppingItem]> {
        ensureRemoteSync(householdId)
        return localStore.observeShoppingItems(householdId: householdId, status: .bought)
    }

    func observeAllItems(householdId: String) -> AsyncStream<[ShoppingItem]> {
        ensureRemoteSync(householdId)
        return localStore.observeShoppingItems(householdId: householdId, status: nil)
    }

    func addItem(householdId: String, item: ShoppingItem) async throws -> String {
        let now = Date()
        let newItem = copy(
            item,
            id: item.id.isEmpty ? UUID().uuidString : item.id,
            normalizedName: normalizeItemName(item.name),
            status: .active,
            createdAt: item.createdAt ?? now,
            updatedAt: now
        )
        localStore.upsertShoppingItem(householdId: householdId, item: newItem)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .item, targetId: newItem.id)
        try await logActivity(householdId: householdId, type: .itemAdded, item: newItem, userId: newItem.addedBy, userName: newItem.addedByName)
        try? await syncQueueProcessor.flush(householdId: householdId)
        return newItem.id
    }

    func updateItem(householdId: String, item: ShoppingItem) async throws {
        let updated = copy(item, normalizedName: normalizeItemName(item.name), updatedAt: Date())
        localStore.upsertShoppingItem(householdId: householdId, item: updated)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .item, targetId: updated.id)
        let actor = await firstCurrentUser()
        try await logActivity(
            householdId: householdId,
            type: .itemUpdated,
            item: updated,
            userId: actor?.id ?? updated.addedBy,
            userName: actor?.displayName ?? updated.addedByName
        )
        try? await syncQueueProcessor.flush(householdId: householdId)
    }

    func markAsBought(householdId: String, itemId: String, userId: String, userName: String) async throws {
        let current = try await getItem(householdId: householdId, itemId: itemId)
        var updated = current
        updated.status = .bought
        updated.boughtBy = userId
        updated.boughtByName = userName
        updated.updatedAt = Date()
        localStore.upsertShoppingItem(householdId: householdId, item: updated)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .item, targetId: updated.id)
        try await logActivity(householdId: householdId, type: .itemBought, item: updated, userId: userId, userName: userName)
        try? await syncQueueProcessor.flush(householdId: householdId)
    }

    func markAsActive(householdId: String, itemId: String) async throws {
        let current = try await getItem(householdId: householdId, itemId: itemId)
        let actor = await firstCurrentUser()
        var updated = current
        updated.status = .active
        updated.boughtBy = nil
        updated.boughtByName = nil
        updated.updatedAt = Date()
        localStore.upsertShoppingItem(householdId: householdId, item: updated)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .item, targetId: updated.id)
        try await logActivity(householdId: householdId, type: .itemRestored, item: updated, userId: actor?.id ?? "", userName: actor?.displayName ?? "")
        try? await syncQueueProcessor.flush(householdId: householdId)
    }

    func deleteItem(householdId: String, itemId: String) async throws {
        let current = localStore.fetchShoppingItem(householdId: householdId, itemId: itemId)
        let actor = await firstCurrentUser()
        localStore.deleteShoppingItem(householdId: householdId, itemId: itemId)
        syncQueueProcessor.enqueueDelete(householdId: householdId, targetType: .item, targetId: itemId)
        if let current {
            try await logActivity(householdId: householdId, type: .itemDeleted, item: current, userId: actor?.id ?? "", userName: actor?.displayName ?? "")
        }
        try? await syncQueueProcessor.flush(householdId: householdId)
    }

    func getItem(householdId: String, itemId: String) async throws -> ShoppingItem {
        if let local = localStore.fetchShoppingItem(householdId: householdId, itemId: itemId) {
            return local
        }
        let remote = try await remoteDataSource.getItem(householdId: householdId, itemId: itemId)
        localStore.upsertShoppingItem(householdId: householdId, item: remote)
        return remote
    }

    func toggleFavorite(householdId: String, itemId: String, isFavorite: Bool) async throws {
        let current = try await getItem(householdId: householdId, itemId: itemId)
        let updated = copy(current, isFavorite: isFavorite, updatedAt: Date())
        localStore.upsertShoppingItem(householdId: householdId, item: updated)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .item, targetId: updated.id)
        try? await syncQueueProcessor.flush(householdId: householdId)
    }

    func flushPendingSync(householdId: String) async throws {
        try await syncQueueProcessor.flush(householdId: householdId)
    }

    func forceRefreshSync(householdId: String) {
        itemListeners.removeValue(forKey: householdId)?.remove()
        registerListener(householdId)
    }

    func clearListeners() {
        itemListeners.values.forEach { $0.remove() }
        itemListeners.removeAll()
    }

    private func ensureRemoteSync(_ householdId: String) {
        guard itemListeners[householdId] == nil else { return }
        registerListener(householdId)
    }

    private func registerListener(_ householdId: String) {
        itemListeners[householdId] = remoteDataSource.listenToItems(householdId: householdId) { [weak self] items in
            Task { @MainActor in
                guard let self else { return }
                self.localStore.mergeRemoteShoppingItems(householdId: householdId, items: items)
                try? await self.syncQueueProcessor.flush(householdId: householdId)
            }
        }
    }

    private func logActivity(householdId: String, type: ActivityType, item: ShoppingItem, userId: String, userName: String) async throws {
        let activity = ActivityLog(
            id: UUID().uuidString,
            householdId: householdId,
            type: type,
            itemId: item.id,
            itemName: item.name,
            actorUserId: userId,
            actorDisplayName: userName,
            createdAt: Date()
        )
        localStore.upsertActivity(activity)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .activity, targetId: activity.id)
    }

    private func firstCurrentUser() async -> UserProfile? {
        for await user in authRepository.observeCurrentUser() { return user }
        return nil
    }

    private func copy(
        _ item: ShoppingItem,
        id: String? = nil,
        normalizedName: String? = nil,
        status: ItemStatus? = nil,
        isFavorite: Bool? = nil,
        createdAt: Date? = nil,
        updatedAt: Date? = nil
    ) -> ShoppingItem {
        var item = item
        if let id { item.id = id }
        if let normalizedName { item.normalizedName = normalizedName }
        if let status { item.status = status }
        if let isFavorite { item.isFavorite = isFavorite }
        if let createdAt { item.createdAt = createdAt }
        if let updatedAt { item.updatedAt = updatedAt }
        return item
    }
}
